### PR TITLE
Optimize JS callbacks

### PR DESCRIPTION
**Pull request by jasongin**
- Change `JSCallbackArgs` from a `class` with heap-allocated args array to a `ref struct` with stack-allocated args span.
 - Fix an issue uncovered by the perf test (due to new timing or memory use patterns I guess?) in which the JS GC could collect the static class object that caused the `JSPropertyDescriptor` GC handle to be finalized even though JS code could still call the method afterward.

With these changes, performance on the edgejs scenario is 10-15% faster.

### CODE REVIEW

File: Runtime/JSValue.cs [hunk 1]
```diff
@@ -10,31 +10,29 @@ namespace NodeApi;
 public readonly struct JSValue : IEquatable<JSValue>
 {
     private readonly napi_value _handle;
-    private readonly JSValueScope? _scope;
-
-    public readonly JSValueScope Scope =>
-        _scope ?? JSValueScope.Current ?? throw new InvalidOperationException("No current scope");
+    private readonly napi_env _scope;
```
Comment (vmoroz) [inline Runtime/JSValue.cs]: >private readonly napi_env _scope; [](http://example.com/codeflow?start=4&length=33)

We effectively trading performance for safety. While it may seem a good approach, it goes against our ability using this framework in server environment.

I do not like this change. After that any programming error will result in memory corruption which will be every expensive or impossible to find. I spent too many weeks catching memory corruptions.

If we want speed, then we must find an alternative, but do not trade safety of the code.

File: Runtime/JSCallbackArgs.cs [hunk 2]
```diff
@@ -1,51 +1,57 @@
+using System;
 using System.Runtime.InteropServices;
 using static NodeApi.JSNativeApi.Interop;
 
 namespace NodeApi;
 
-public class JSCallbackArgs
+public readonly ref struct JSCallbackArgs
```
Comment (vmoroz) [inline Runtime/JSCallbackArgs.cs:7]: >readonly ref struct [](http://example.com/codeflow?start=7&length=19)

This is a good change! I can see how it makes the perf 15% faster because we effectively avoid extra memory allocation per each call.

File: Runtime/JSCallbackArgs.cs [hunk 3]
```diff
@@ -1,51 +1,57 @@
...
 {
-    private readonly JSValue[] _args;
+    private readonly Span<JSValue> _args;
```
Comment (vmoroz) [inline Runtime/JSCallbackArgs.cs]: >Span [](http://example.com/codeflow?start=21&length=4)

Should it be ReadOnlySpan?

File: Runtime/JSCallbackArgs.cs (hunk 2 above)
Comment (vmoroz) [inline Runtime/JSCallbackArgs.cs:7]: >ref struct  [](http://example.com/codeflow?start=16&length=11)

Can we use 'struct' instead of 'ref struct'?

**Comment by vmoroz** [comment]
Please do not merge this PR until we can find a better solution. Improving the performance for each interop call is very important, but we must not regress on the code safety.

File: Runtime/JSCallbackArgs.cs (hunk 2 above)
Comment (jasongin) [reply Runtime/JSCallbackArgs.cs:7]: It has to be `ref struct` so that it can have a `Span` member, to support stack-allocation of the args array.

File: Runtime/JSCallbackArgs.cs (hunk 3 above)
Comment (jasongin) [reply Runtime/JSCallbackArgs.cs]: Yes.

**Comment by vmoroz** [comment]
Please see the fix that keeps `JSValue` safe and still avoids memory allocation for argument passing: https://github.com/jasongin/napi-dotnet/commit/f56929f8437918fc14620a6788de5b212b61fe21

File: Runtime/JSValue.cs (hunk 1 above)
Comment (jasongin) [reply Runtime/JSValue.cs]: I agree memory safety is important.

I think I was confusing the `napi_env` and `napi_handle_scope` in this change (because they're sort of combined now in `JSValueScope`) and assuming a closed scope would result in a `napi_invalid_arg` status from any calls using handles from that scope. But the `napi_env` is still valid, and if the `napi_value` is literally just a pointer, then it could be dereferenced after its scope was closed and memory was freed.

I'll try to find a better solution.

Also we'll need some tests for handle scopes including negative tests. I see there are [some node-addon-api handlescope tests](https://github.com/nodejs/node-addon-api/blob/main/test/handlescope.cc) that could be ported over, though it doesn't look like that includes negative tests.

**Comment by jasongin** [comment]
>Please see the fix that keeps JSValue safe and still avoids memory allocation for argument passing: https://github.com/jasongin/napi-dotnet/commit/f56929f8437918fc14620a6788de5b212b61fe21

Oh thanks! That looks good. I'll merge it into this PR.

**Comment by vmoroz** [comment]
Just a small simplification to put more logic in the `JSCallbackArgs`: https://github.com/jasongin/napi-dotnet/commit/2cd3472d95cc37f092551037f65d2b70dd12bf30

**Review by vmoroz — APPROVED** [review]
LGTM
